### PR TITLE
OrderValidationUtils: exclude dust orders

### LINKED ISSUE #2162: OrderValidationUtils.sol: Changes for v3
This issue represents the work to be done in [the Asana item "DevUtils upgrade - and deploy with Staking"](https://app.asana.com/0/1139304696829186/1139688593501305/f).

Theo said: "Functions to focus on: `GetOrderRelevantStates`, `GetEthBalances`"

@abandeali1 said: "account for the ordering of transfers when calculating balances. For example, a maker’s balance should account for the taker -> maker transfer that happens before the maker -> taker transfer. Fee balances should also account for both the taker -> maker transfer and the maker -> taker transfers that happen before fees are paid."

### PR DESCRIPTION
**Pull request by feuGeneA**

### CODE REVIEW

File: contracts/dev-utils/test/order_validation_utils.ts [hunk 1]
```diff
@@ -412,6 +412,172 @@ describe('OrderValidationUtils/OrderTransferSimulatorUtils', () => {
...
+            const [, fillableTakerAssetAmount] = await devUtils.getOrderRelevantState.callAsync(
+                signedOrder,
+                signedOrder.signature,
+            );
+
+            // Assert.
+
+            expect(fillableTakerAssetAmount).to.bignumber.equal(0);
+        });
+        it('should return correct info even when there are no fees specified', async () => {
```
Comment (albrow) [inline contracts/dev-utils/test/order_validation_utils.ts]: I'm not qualified to review the Solidity code but these tests look really good 😄 Looks like it solves the issue we were running into while working on Mesh and additionally brings further improvements to test coverage.

File: contracts/dev-utils/contracts/src/OrderValidationUtils.sol [hunk 2]
```diff
@@ -73,36 +73,65 @@ contract OrderValidationUtils is
...
-                    order.makerAssetAmount,
+                    makerAssetAmount.safeAdd(makerFee),
                     takerAssetAmount
                 );
-
-            // If `makerAsset` does not equal `makerFeeAsset`, the % that can be filled is the lower of
-            // (transferableMakerAssetAmount / makerAssetAmount) and (transferableMakerAssetFeeAmount / makerFee)
+            } else if (order.makerFeeAssetData.equals(order.takerAssetData)) {
+                // The % that can be filled is
+                // (transferableMakerAssetAmount + takerAssetAmount) / (makerAssetAmount + makerFee)
```
Comment (abandeali1) [inline contracts/dev-utils/contracts/src/OrderValidationUtils.sol]: This math looks off to me. Let's sync offline - lots of edge cases here.

File: contracts/dev-utils/contracts/src/OrderValidationUtils.sol [hunk 3]
```diff
@@ -73,36 +73,65 @@ contract OrderValidationUtils is
...
+                && multiAssetContainsAsset(order.takerAssetData, order.makerFeeAssetData)
+            ) {
+                // find the asset data in order.takerAssetData which matches the makerFeeAsset
+                (
+                    ,
+                    uint256[] memory amounts,
+                    bytes[] memory nestedAssetData
+                ) = decodeMultiAssetData(order.takerAssetData);
+
+                for (uint i=0; i < amounts.length; i++) {
```
Comment (abandeali1) [inline contracts/dev-utils/contracts/src/OrderValidationUtils.sol]: `uint256 i = 0`

File: contracts/dev-utils/contracts/src/OrderValidationUtils.sol [hunk 4]
```diff
@@ -73,36 +73,65 @@ contract OrderValidationUtils is
...
+                    if (order.makerFeeAssetData.equals(nestedAssetData[i])) {
```
Comment (abandeali1) [inline contracts/dev-utils/contracts/src/OrderValidationUtils.sol]: Let's create a helper function that compares the first 4 bytes of each `assetData`. We do this a lot throughout these contracts.

File: contracts/dev-utils/contracts/src/OrderValidationUtils.sol [hunk 5]
```diff
@@ -73,36 +73,65 @@ contract OrderValidationUtils is
...
+                        transferableTakerAssetAmount = LibMath.getPartialAmountFloor(
```
Comment (abandeali1) [inline contracts/dev-utils/contracts/src/OrderValidationUtils.sol]: I believe this math is off too.

**Pull request closed without merging**